File: frontend/src/types/customer.ts
/** Mirrors the backend Pydantic `Read` schemas in app/schemas/customer.py. */

export type CustomerStatus = "active" | "archived";
export type ContactKind = "phone" | "email" | "address" | "other";
export type InteractionKind = "call" | "email" | "meeting" | "chat" | "other";

export const CONTACT_KINDS: ContactKind[] = ["phone", "email", "address", "other"];
export const INTERACTION_KINDS: InteractionKind[] = [
  "call",
  "email",
  "meeting",
  "chat",
  "other",
];

export interface Page<T> {
  items: T[];
  total: number;
}

export interface Customer {
  id: string;
  display_name: string;
  company: string | null;
  status: CustomerStatus;
  archived_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface CustomerDetail extends Customer {
  contacts: ContactDetail[];
}

export interface CustomerCreate {
  display_name: string;
  company?: string | null;
}

export interface CustomerUpdate {
  display_name?: string;
  company?: string | null;
  status?: CustomerStatus;
}

export interface ContactDetail {
  id: string;
  customer_id: string;
  kind: ContactKind;
  value: string;
  label: string | null;
  is_primary: boolean;
  created_at: string;
}

export interface ContactDetailInput {
  kind: ContactKind;
  value: string;
  label?: string | null;
  is_primary?: boolean;
}

export interface Interaction {
  id: string;
  customer_id: string;
  kind: InteractionKind;
  subject: string | null;
  body: string;
  occurred_at: string;
  author: string | null;
  created_at: string;
}

export interface InteractionInput {
  kind: InteractionKind;
  subject?: string | null;
  body?: string;
  occurred_at: string;
  author?: string | null;
}

export interface Note {
  id: string;
  customer_id: string;
  body: string;
  created_at: string;
  updated_at: string;
}

export interface Attachment {
  id: string;
  customer_id: string;
  note_id: string | null;
  filename: string;
  content_type: string;
  size_bytes: number;
  created_at: string;
}

File: backend/app/schemas/customer.py
"""Pydantic v2 schemas for the customer-management API."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Annotated, Generic, Literal, TypeVar

from pydantic import BaseModel, ConfigDict, Field

CustomerStatus = Literal["active", "archived"]
ContactKind = Literal["phone", "email", "address", "other"]
InteractionKind = Literal["call", "email", "meeting", "chat", "other"]

NonEmptyStr = Annotated[str, Field(min_length=1)]

T = TypeVar("T")


class Page(BaseModel, Generic[T]):
    items: list[T]
    total: int


class ListQuery(BaseModel):
    """Query params for the customer list; invalid values yield 422."""

    q: str | None = None
    status: CustomerStatus | None = None
    limit: int = Field(default=20, ge=1, le=100)
    offset: int = Field(default=0, ge=0)


# --------------------------------------------------------------------------- #
# Customer
# --------------------------------------------------------------------------- #
class CustomerBase(BaseModel):
    display_name: NonEmptyStr
    company: str | None = None


class CustomerCreate(CustomerBase):
    pass


class CustomerUpdate(BaseModel):
    display_name: NonEmptyStr | None = None
    company: str | None = None
    status: CustomerStatus | None = None


class CustomerRead(CustomerBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    status: CustomerStatus
    archived_at: datetime | None
    created_at: datetime
    updated_at: datetime


class CustomerDetailRead(CustomerRead):
    """Customer plus eagerly loaded contacts (detail endpoint only)."""

    contacts: list[ContactDetailRead] = []


# --------------------------------------------------------------------------- #
# Contact details
# --------------------------------------------------------------------------- #
class ContactDetailBase(BaseModel):
    kind: ContactKind
    value: NonEmptyStr
    label: str | None = None
    is_primary: bool = False


class ContactDetailCreate(ContactDetailBase):
    pass


class ContactDetailUpdate(BaseModel):
    kind: ContactKind | None = None
    value: NonEmptyStr | None = None
    label: str | None = None
    is_primary: bool | None = None


class ContactDetailRead(ContactDetailBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    customer_id: uuid.UUID
    created_at: datetime


# --------------------------------------------------------------------------- #
# Interactions
# --------------------------------------------------------------------------- #
class InteractionBase(BaseModel):
    kind: InteractionKind
    subject: str | None = None
    body: str = ""
    occurred_at: datetime
    author: str | None = None


class InteractionCreate(InteractionBase):
    pass


class InteractionUpdate(BaseModel):
    kind: InteractionKind | None = None
    subject: str | None = None
    body: str | None = None
    occurred_at: datetime | None = None
    author: str | None = None


class InteractionRead(InteractionBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    customer_id: uuid.UUID
    created_at: datetime


# --------------------------------------------------------------------------- #
# Notes
# --------------------------------------------------------------------------- #
class NoteBase(BaseModel):
    body: NonEmptyStr


class NoteCreate(NoteBase):
    pass


class NoteUpdate(BaseModel):
    body: NonEmptyStr | None = None


class NoteRead(NoteBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    customer_id: uuid.UUID
    created_at: datetime
    updated_at: datetime


# --------------------------------------------------------------------------- #
# Attachments
# --------------------------------------------------------------------------- #
class AttachmentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    customer_id: uuid.UUID
    note_id: uuid.UUID | None
    filename: str
    content_type: str
    size_bytes: int
    created_at: datetime


CustomerDetailRead.model_rebuild()
